akka-http-core/src/main/scala/akka/http/engine/ws/FrameOutHandler.scala

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (C) 2009-2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+
+package akka.http.engine.ws
+
+import scala.concurrent.duration.FiniteDuration
+
+import akka.stream.stage._
+import akka.http.util.Timestamp
+import FrameHandler.{ UserHandlerCompleted, ActivelyCloseWithCode, PeerClosed, DirectAnswer }
+import Websocket.Tick
+
+/**
+ * Implements the transport connection close handling at the end of the pipeline.
+ *
+ * INTERNAL API
+ */
+private[http] class FrameOutHandler(serverSide: Boolean, _closeTimeout: FiniteDuration) extends StatefulStage[AnyRef, FrameStart] {
+  def initial: StageState[AnyRef, FrameStart] = Idle
+  def closeTimeout: Timestamp = Timestamp.now + _closeTimeout
+
+  object Idle extends CompletionHandlingState {
+    def onPush(elem: AnyRef, ctx: Context[FrameStart]): SyncDirective = elem match {
+      case start: FrameStart   ⇒ ctx.push(start)
+      case DirectAnswer(frame) ⇒ ctx.push(frame)
+      case PeerClosed(code, reason) if !code.exists(Protocol.CloseCodes.isError) ⇒
+        // let user complete it, FIXME: maybe make configurable? immediately, or timeout
```

Comment:
The [spec says](http://tools.ietf.org/html/rfc6455#section-5.5.1):

> If an endpoint receives a Close frame and did not previously send a
   Close frame, the endpoint MUST send a Close frame in response.  (When
   sending a Close frame in response, the endpoint typically echos the
   status code it received.)  It SHOULD do so as soon as practical.  An
   endpoint MAY delay sending a Close frame until its current message is
   sent (for instance, if the majority of a fragmented message is
   already sent, an endpoint MAY send the remaining fragments before
   sending a Close frame).

We currently, fully support half-closed connections, so we wait for the user-handler to complete and only then send our close frame. "as soon as practical" is currently implemented as "as soon as the user-handler deems right". Is this enough or do we need automatic closing (after a timeout?) WDYT?